goonchat linkify has been vastly improved

optimumtact opened the pull request:
You must now explicitly specify if you want linkification by setting
    linkify=TRUE in the to_chat call, right now this is only enabled by
    default for OOC

    The linkification process is now also restricted to only the content
    inside spans that contain the class .linkify. This is because if we ran
    the linkify unrestricted on the entire message we would break things that
    are using URL's for specific purposes.

    For example, goonchat images are served as an <img src="url"> link, and if
    we were to replace the url in that with a clickable a href we would break
    the images being sent to chat

    The combination of flag and span class is entirely for performance, if there's no
    reason to run the linkify process over the message text, we can simply
    avoid doing so and not waste time doing work we don't need to do

Why:
previously in the image below, my name would have appeared as a clickable link in ic messages as well
![](https://file.house/_use.png)


The regex will match links of the form
www.tgstation13.org
https://tgstation13.org
https://tgstation13.org/some/url/shit
http://tgstation13.org
http://www.tgstation13.org
byond://sybil.tgstation13.org 

etc etc

Fixes #33996
Fixes #36924


I added a basic debug console that calls eval, which helps when doing on sight debugging of issues. It also gives you somewhere to send debug output that isn't just the main game screen. So you can leave it in as reuseable for future devs.
![](https://file.house/Aepq.png)

AnturK commented on code/modules/goonchat/browserOutput.dm:
```diff
@@ -227,8 +235,10 @@ GLOBAL_DATUM_INIT(iconCache, /savefile, new("data/iconCache.sav")) //Cache of ic
+		// url_encode the message TWICE (list2params does it once), this way any UTF-8 characters are
+		//able to be decoded by the Javascript.
+		var/list/final = list(url_encode(message), flag)
+		C << output(list2params(final), "browseroutput:output")
```
Was passing other arguments simply unused before ? I don't see any change about accepting them

praisenarsie commented:
fun fact i've reported this as an issue before but cyberboss brushed it off as a feature (not trying to hate on cyber)

optimumtact commented:
okay this seems to be breaking goonchat when you disconnect and reconnect, not sure why, having a look at that code now

optimumtact commented:
I'm not able to see why this is happening! Disconnecting from my dev machine and then reconnecting to a session in progress makes the fancy chat appear broken.

It seems like byond actually removes the UI element, which was very confusing to me

optimumtact commented:
I was able to replicate the bug on master, so I don't believe it's my code at this time, but something to do with my environment.

test merge please

AnturK commented:
#34112

optimumtact commented:
shit, they're probably inside the spans

I'll figure something out for this

MrStonedOne commented:
This made admins ooc messages normal, and player ooc messages bold.

Jazaen commented:
And also it sets player ooc to black, and makes it unchangeable.

BeeSting12 commented:
Not a huge fan of admin OOC being normal text over bolded. https://file.house/e8n5.png

optimumtact commented:
feature

AnturK commented:
How's this going.

optimumtact commented:
I've been ignoring it for a while, will try to get back to it this week

optimumtact commented:
I was lying to myself and you

optimumtact commented:
todo, don't break asset serving
test it fixes #36924 
remove the changes I made to try to get firebug working

optimumtact commented:
clearly I will never finish this goddamn change

optimumtact committed:
goonchat linkify has been vastly improved

You must now explicitly specify if you want linkification by setting
linkify=TRUE in the to_chat call, right now this is only enabled by
default for OOC

The linkification process is now also restricted to only the content
inside spans that contain the class .linkify. This is because if we ran
the linkify unrestricted on the entire message we would break things that
are using URL's for specific purposes.

For example, goonchat images are served as an <img src="url"> link, and if
we were to replace the url in that with a clickable a href we would break
the images being sent to chat

The combination of flag and span class is entirely for performance, if there's no
reason to run the linkify process over the message text, we can simply
avoid doing so and not waste time doing work we don't need to do

optimumtact committed:
Adds a ghetto console for testing

Admin debug verb that enables it, consisting of a text area, several
buttons to set debug verbosity and to run simple scripts from the text
area (passed directly to eval)

optimumtact committed:
Fix the admin messages not being bolded

We were skipping the outside div span element by using html instead of
outerHTML

Also there was some ordering issues with the span and font end tags on the
admin ooc message

optimumtact commented:
This is ready for a test merge @tgstation/commit-access 

I don't want to merge it without at least one test for compatibility on peoples browsers.

PKPenguin321 commented:
Hard work and perseverance
Good job oranges

optimumtact commented on code/modules/goonchat/browserassets/js/browserOutput.js:
```diff
@@ -6,23 +6,13 @@
-var triggerError = attachErrorHandler('chatDebug', true);
 var escaper = encodeURIComponent || escape;
 var decoder = decodeURIComponent || unescape;
-window.onerror = function(msg, url, line, col, error) {
```
@MrStonedOne I don't think this is used, but do you have any use for it?

optimumtact replied:
herp, I'll fix that

optimumtact commented on code/modules/goonchat/browserassets/js/browserOutput.js:
this should be a global

pigeonsk replied:
We have a global but it's outdated and barely used
https://github.com/tgstation/tgstation/blob/4f7a382cdd4e4d0b2fda1784b59f7dee3e32e15d/code/_globalvars/regexes.dm#L4

optimumtact replied:
that's also in DM, which isn't very helpful for this javascript, this was just a note to myself to move the regex up to the top level and init it in the goonchat startup function

pigeonsk replied:
whoops forgot it was js

DaxDupont commented:
![pruwbzr 1](https://user-images.githubusercontent.com/17237624/39268188-6912cb16-48cf-11e8-9870-4c60102858a1.png)
can you make this not a seperate tab and toss it under debug please?
rid: 87020

DaxDupont commented:
![5htqspu 1](https://user-images.githubusercontent.com/17237624/39272660-d20b4672-48dc-11e8-9ad3-3f76b595772a.png)
these are no longer links where they used to be.
87023

optimumtact committed:
Add changes to some message types

DaxDupont commented:
Links in deadchat are no longer linkified round: 87192

optimumtact commented:
on it

optimumtact committed:
Deadchat now linkifies text

optimumtact commented:
Added dchat to the list of messages that get linkified
now it's
ooc
dchat
and admin_messages

optimumtact commented on code/modules/goonchat/browserOutput.dm:
```diff
@@ -139,6 +139,12 @@ GLOBAL_DATUM_INIT(iconCache, /savefile, new("data/iconCache.sav")) //Cache of ic
 	if(volume)
 		adminMusicVolume = CLAMP(text2num(volume), 0, 100)
 
+//Don't call this, it will simply lock your client up and never enable firebug!
```
update this it's not true

DaxDupont commented:
87282 dead chat is blank
![dt1xpj3 1](https://user-images.githubusercontent.com/17237624/39493885-b28a43f4-4d94-11e8-8565-93198657c84a.png)
only runtimes are
```
[20:59:52] Runtime in core.dm, line 23: list index out of bounds
proc name: SelectTgsApi (/world/proc/SelectTgsApi)
src: world
call stack:
world: SelectTgsApi("3.2.2.0")
world: TgsNew(null)
world: New()
```
```
[20:59:52] Runtime in core.dm, line 10: Cannot create objects of type null.
proc name: TgsNew (/world/TgsNew)
src: world
call stack:
world: TgsNew(null)
world: New()
```

optimumtact commented:
I will be finishing this this weekend, just need to fix the deadchat issue and go through adminpm and adminhelp so that they show links

AnturK commented:
This going anywhere ?

optimumtact commented:
if you've looked at adminpm adn adminhelp you would know why I'm struggling to finish this.

will try to do it this week though

KorPhaeron commented:
Reopen when you are ready
